### ISSUE TITLE
Bug when fetching list of products when the product has optional values in it

### ISSUE THREAD

**Issue by afsalthaj** [zio/zio-config#193]
This is applicable in `zio-config-typesafe`
There is an issue when fetching list of case classes, such that if one of the field in the case class is None, reading the config gives wrong results. See the below example

```scala
final case class Aws(region: Option[String])
val desc = description[Aws]

object Main extends App {
  val hocconSourceList =
    hocon(
      Right(
        """
        mylist = [{
         region: abc
        },
        {
         region: null
        }]
       """
      )
    )

  val runtime = new DefaultRuntime {}
     
   runtime.unsafeRun(
     read(nested("mylist")(list(Aws.desc)) from hocconSourceList).either
   )

// Right(List(Aws(None)) (expected: Right(List(Aws("abc")))
}


```

**Comment by afsalthaj** [comment]
This is mainly due to the fact that typesafe config implementation fail early when trying to fetch a list of values after identifying the valueType is a list. It fails, when it hits a `null` which wasn't handled. 

This indicate that fetching a value for a key need to return ::[Option[A]] and not necessarily ::[A]. The option is required to indicate that one of them is `null` (In the above example region is  `null`, and we can't discard this.). In other words, we still need to continue without failing early and still need to return non-failed A in ZIO[E, A]. 

I couldn't get much time investigating this, however, it may need some thinking towards implementing ReadFunctions.

**Comment by afsalthaj** [comment]
An initial version is here:
https://github.com/zio/zio-config/blob/3876998d79731955bf01fce783a13ee30acf25c5/examples/src/main/scala/zio/config/examples/typesafe/NullAndOptionalConfig.scala

**Comment by afsalthaj** [comment]
#197

**Comment by afsalthaj** [comment]
Fixed in #236
